pkg/koordlet/resmanager/cpu_suppress.go

```diff
@@ -158,98 +159,102 @@ func (r *CPUSuppress) calculateBESuppressCPU(node *corev1.Node, nodeMetric *metr
-
-	// initially sorted by processor id
-	prioritizedCPUs := nodeCPUInfo.ProcessorInfos
-	sort.Slice(prioritizedCPUs, func(i, j int) bool {
-		return prioritizedCPUs[i].CPUID < prioritizedCPUs[j].CPUID
-	})
-
 	// getNodeIndex is a function to calculate an index for every numa node or socket
 	getNodeIndex := func(info util.ProcessorInfo) int32 {
+		// (nodeId, socketId) => nodeIndex
 		return (info.NodeID + numProcessors) * (info.SocketID + 1)
 	}
+	cpuBucketOfNode := map[int32][]util.ProcessorInfo{}
+	for _, p := range processorInfos {
+		nodeIndex := getNodeIndex(p)
+		cpuBucketOfNode[nodeIndex] = append(cpuBucketOfNode[nodeIndex], p)
+	}
 
-	// get the relative indexes of the processor on its node
-	indexOfNode := map[int32]int32{}
-	numProcessorsOfNode := map[int32]int32{}
-	for _, processor := range prioritizedCPUs {
-		numProcessorsOfNode[getNodeIndex(processor)]++
-		if indexOfNode[processor.CoreID] <= 0 {
-			indexOfNode[processor.CoreID] = numProcessorsOfNode[getNodeIndex(processor)]
-		}
+	// change cpuBucket map to array
+	cpuBucket := [][]util.ProcessorInfo{}
+	for _, processorInfos := range cpuBucketOfNode {
+		cpuBucket = append(cpuBucket, processorInfos)
 	}
 
-	// first we prefer the scattered node/socket, secondly the same core
-	// e.g.
-	// for a processorInfo/prioritizedCPUs list:
-	//	CPUID	CoreID	SocketID	NodeID
-	//	0		0		0			0
-	//	1		1		0			0
-	//	2		2		1			1
-	//	3		3		1			1
-	//	4		0		0			0
-	//	5		1		0			0
-	//	6		2		1			1
-	//	7		3		1			1
-	// it will get sorted as below:
-	//	CPUID	CoreID	SocketID	NodeID
-	//	7		3		1			1
-	//	3		3		1			1
-	//	5		1		0			0
-	//	1		1		0			0
-	//	6		2		1			1
-	//	2		2		1			1
-	//	4		0		0			0
-	//	0		0		0			0
-	sort.Slice(prioritizedCPUs, func(i, j int) bool {
-		a, b := prioritizedCPUs[i], prioritizedCPUs[j]
-		if indexOfNode[a.CoreID] != indexOfNode[b.CoreID] {
-			return indexOfNode[a.CoreID] > indexOfNode[b.CoreID]
-		}
-		if numProcessorsOfNode[getNodeIndex(a)] != numProcessorsOfNode[getNodeIndex(b)] {
-			return numProcessorsOfNode[getNodeIndex(a)] > numProcessorsOfNode[getNodeIndex(b)]
-		}
-		if a.CoreID != b.CoreID {
-			return a.CoreID > b.CoreID
+	for index := range cpuBucket {
+		sort.Slice(cpuBucket[index], func(i, j int) bool {
+			if cpuBucket[index][i].CoreID == cpuBucket[index][j].CoreID {
+				return cpuBucket[index][i].CPUID < cpuBucket[index][j].CPUID
+			}
+			return cpuBucket[index][i].CoreID < cpuBucket[index][j].CoreID
+		})
+	}
+
+	sort.Slice(cpuBucket, func(i, j int) bool {
+		if len(cpuBucket[i]) == len(cpuBucket[j]) {
+			return cpuBucket[i][0].CPUID < cpuBucket[j][0].CPUID
 		}
-		return a.CPUID > b.CPUID
+		return len(cpuBucket[i]) > len(cpuBucket[j])
 	})
 
 	needCPUs := cpus
-	for i := range prioritizedCPUs {
+	usedCpu := map[int32]bool{}
+	// select same core cpu id
+	preNeedCpus := int32(-1)
+	i := 0
+	for ; i < len(cpuBucket); i = (i + 1) % len(cpuBucket) {
+		if needCPUs <= 1 {
+			break
+		}
+		if i == len(cpuBucket) {
```

Comment:
this never happens due to `(i + 1) % len(cpuBucket)`
consider check then current and pre length of usedCpu